usersession/userd/launcher.go

```diff
@@ -123,6 +131,105 @@ func (s *Launcher) OpenURL(addr string, sender dbus.Sender) *dbus.Error {
 	return nil
 }
 
+// OpenDesktopEntry implements the 'OpenDesktopEntry' method of the 'io.snapcraft.Launcher'
+// DBus interface. Before the provided desktop_file is parsed it is validated against a list
+// of allowed locations.
+func (s *Launcher) OpenDesktopEntry(desktop_file_id string, sender dbus.Sender) *dbus.Error {
+
+	return s.OpenDesktopEntryEnv(desktop_file_id, []string{}, sender)
+}
+
+// OpenDesktopEntryEnv implements the 'OpenDesktopEntryEnv' method of the 'io.snapcraft.Launcher'
+// DBus interface.
+func (s *Launcher) OpenDesktopEntryEnv(desktop_file_id string, env []string, sender dbus.Sender) *dbus.Error {
+	exec_command, err := s.readExecCommandFromDesktopFile(s.desktopFileIdToFilename(desktop_file_id))
+
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	// Passing exec variables between confined snaps raises unanswered questions and they are not required
+	// for the simple cases.  For now, we don't have support for passing them in the dbus API and truncate
+	// the exec command at the first exec variable.
+	// https://standards.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#exec-variables
+	exec_command = strings.SplitN(exec_command, "%", 2)[0]
+
+	args := strings.Split(exec_command, " ")
+	cmd := exec.Command(args[0], args[1:]...)
+
+	cmd.Env = os.Environ()
+	for _, e := range env {
+		cmd.Env = append(cmd.Env, e)
+	}
+
+	if err := cmd.Start(); err != nil {
+		return dbus.MakeFailedError(fmt.Errorf("cannot run %q", exec_command))
+	}
+
+	return nil
+}
+
+// desktopFileIdToFilename determines the path associated with a desktop file ID.
+func (s *Launcher) desktopFileIdToFilename(desktop_file_id string) string {
+	splitFileId := strings.Split(desktop_file_id, "-")
+
+	var desktop_file string
+
+	for _, dir := range strings.Split(os.Getenv("XDG_DATA_DIRS"), ":") {
+		var fileStat os.FileInfo
+
+		for i := 0; i != len(splitFileId); i = i + 1 {
+			desktop_file = dir + "/applications/" + strings.Join(splitFileId[0:i], "/") + "/" + strings.Join(splitFileId[i:], "-")
```

Comment:
Use `filepath.Join` to construct new paths (from the `path/filepath` standard library package).

Also, would this code correctly locate a desktop file named `/usr/share/applications/foo-bar/baz-qux.desktop`?  How does it compare to what e.g. GLib does?